### Load the libraries

In [8]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import seaborn as sns
import matplotlib.pyplot as plt
sns.set()

### Load the train dataset

In [9]:
raw_data = pd.read_csv("Binary_predictors.csv")
raw_data

SAT Admitted  Gender
0    1363       No    Male
1    1792      Yes  Female
2    1954      Yes  Female
3    1653       No    Male
4    1593       No    Male
..    ...      ...     ...
163  1722      Yes  Female
164  1750      Yes    Male
165  1555       No    Male
166  1524       No    Male
167  1461       No    Male

[168 rows x 3 columns]

Let's change the categorical values to 0s and 1s. 

In [10]:
# Map the train dataset
data = raw_data
data['Admitted'] = data['Admitted'].map({'Yes': 1, 'No': 0})
data['Gender'] = data['Gender'].map({'Female': 1, 'Male': 0})
data

SAT  Admitted  Gender
0    1363         0       0
1    1792         1       1
2    1954         1       1
3    1653         0       0
4    1593         0       0
..    ...       ...     ...
163  1722         1       1
164  1750         1       0
165  1555         0       0
166  1524         0       0
167  1461         0       0

[168 rows x 3 columns]

### Assign the dependent and independent variables

In [11]:
y = data['Admitted']
x1 = data[['SAT','Gender']]

In [12]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()

# Get the regression summary
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.120117
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      165
Method:                           MLE   Df Model:                            2
Date:                Fri, 15 Jul 2022   Pseudo R-squ.:                  0.8249
Time:                        15:02:24   Log-Likelihood:                -20.180
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.118e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -68.3489     16.454     -4.154      0.000    -100.598     -36.100
SAT            0.0406      0.010      4.129      0.000       0.021       0.060
Gender         1.9449      0.846      2.299      0.022       0.287       3.603
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

The LLR p value ~0.000, and so the model is significant. The variable gender and SAT are also significant (small p-values).Using the logit equation, after calculating the odds of a female getting addmitted to the odds of a male getting admitted, a simplified equation was created, for which we will not calculate the exponential value of 1.9449. 

In [13]:
np.exp(1.9449)

6.992932526814459

### Calculate the accuracy of the model

In [14]:
# This is a method to change the formatting of np arrays
np.set_printoptions(formatter={'float': lambda x: "{0:0.2f}".format(x)})

results_log.predict()

array([0.00, 1.00, 1.00, 0.23, 0.02, 0.99, 1.00, 1.00, 1.00, 0.01, 1.00,
       1.00, 0.76, 0.00, 0.60, 1.00, 0.11, 0.12, 0.51, 1.00, 1.00, 1.00,
       0.00, 0.01, 0.97, 1.00, 0.48, 0.99, 1.00, 0.99, 0.00, 0.83, 0.25,
       1.00, 1.00, 1.00, 0.31, 1.00, 0.23, 0.00, 0.02, 0.45, 1.00, 0.00,
       0.99, 0.00, 0.99, 0.00, 0.00, 0.01, 0.00, 1.00, 0.92, 0.02, 1.00,
       0.00, 0.37, 0.98, 0.12, 1.00, 0.00, 0.78, 1.00, 1.00, 0.98, 0.00,
       0.00, 0.00, 1.00, 0.00, 0.78, 0.12, 0.00, 0.99, 1.00, 1.00, 0.00,
       0.30, 1.00, 1.00, 0.00, 1.00, 1.00, 0.85, 1.00, 1.00, 0.00, 1.00,
       1.00, 0.89, 0.83, 0.00, 0.98, 0.97, 0.00, 1.00, 1.00, 0.03, 0.99,
       0.96, 1.00, 0.00, 1.00, 0.01, 0.01, 1.00, 1.00, 1.00, 0.00, 0.00,
       0.02, 0.33, 0.00, 1.00, 0.09, 0.00, 0.97, 0.00, 0.75, 1.00, 1.00,
       0.01, 0.01, 0.00, 1.00, 0.00, 0.99, 0.57, 0.54, 0.87, 0.83, 0.00,
       1.00, 0.00, 0.00, 0.00, 1.00, 0.04, 0.00, 0.01, 1.00, 0.99, 0.52,
       1.00, 1.00, 0.05, 0.00, 0.00, 0.00, 0.68, 1.

In [15]:
# An array containing the TRUE (actual) values
np.array(data['Admitted'])

array([0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

In [16]:
# A prediction table (confusion matrix) showing the 
results_log.pred_table()

array([[69.00, 5.00],
       [4.00, 90.00]])

In [17]:
# Some neat formatting to read the table (better when seeing it for the first time)
cm_df = pd.DataFrame(results_log.pred_table())
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index={0: 'Actual 0',1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0         69.0          5.0
Actual 1          4.0         90.0

In [18]:
# Create an array (so it is easier to calculate the accuracy)
cm = np.array(cm_df)
# Calculate the accuracy of the model
accuracy_train = (cm[0,0]+cm[1,1])/cm.sum()
accuracy_train

0.9464285714285714

The accuracy of the model with the train data is 94.6%. The dataset was initially divided into two, with a 90:10 ratio, as train and test datasets. This was done to resolve the problem of overfitting. Now, we will test the model with the test dataset. 

### Testing the model and assessing its accuracy

In [19]:
# Load the test dataset
test = pd.read_csv('Test_dataset.csv')
test

SAT Admitted  Gender
0   1323       No    Male
1   1725      Yes  Female
2   1762      Yes  Female
3   1777      Yes    Male
4   1665       No    Male
5   1556      Yes  Female
6   1731      Yes  Female
7   1809      Yes  Female
8   1930      Yes  Female
9   1708      Yes    Male
10  1970      Yes  Female
11  2039      Yes  Female
12  1681      Yes  Female
13  1444       No    Male
14  1726       No    Male
15  1876      Yes  Female
16  1485       No    Male
17  1500       No  Female
18  1900      Yes    Male

In [20]:
# Map the test data, as done with the train data
test['Admitted'] = test['Admitted'].map({'Yes': 1, 'No': 0})
test['Gender'] = test['Gender'].map({'Female': 1, 'Male': 0})
test

SAT  Admitted  Gender
0   1323         0       0
1   1725         1       1
2   1762         1       1
3   1777         1       0
4   1665         0       0
5   1556         1       1
6   1731         1       1
7   1809         1       1
8   1930         1       1
9   1708         1       0
10  1970         1       1
11  2039         1       1
12  1681         1       1
13  1444         0       0
14  1726         0       0
15  1876         1       1
16  1485         0       0
17  1500         0       1
18  1900         1       0

In [21]:
# Let's see how the inputs from our x variable look
x

const   SAT  Gender
0      1.0  1363       0
1      1.0  1792       1
2      1.0  1954       1
3      1.0  1653       0
4      1.0  1593       0
..     ...   ...     ...
163    1.0  1722       1
164    1.0  1750       0
165    1.0  1555       0
166    1.0  1524       0
167    1.0  1461       0

[168 rows x 3 columns]

We see that there are 3 columns- const, SAT, and gender, in that order. Order is very important because the co-efficients of regression will expect it. If we fail to deliver in the correct order, the prediction will be wrong. 

In [22]:
# Create a new variable called test_actual to put in the admission information (the actual outcomes that were observed)
test_actual = test['Admitted']

In [23]:
# Create another variable called test_data which will contain everything from the test dataset except the admission information
test_data = test.drop(['Admitted'],axis=1)

# Use the add_constant from statsmodels to add a constant to the variable
test_data = sm.add_constant(test_data)
test_data

const   SAT  Gender
0     1.0  1323       0
1     1.0  1725       1
2     1.0  1762       1
3     1.0  1777       0
4     1.0  1665       0
5     1.0  1556       1
6     1.0  1731       1
7     1.0  1809       1
8     1.0  1930       1
9     1.0  1708       0
10    1.0  1970       1
11    1.0  2039       1
12    1.0  1681       1
13    1.0  1444       0
14    1.0  1726       0
15    1.0  1876       1
16    1.0  1485       0
17    1.0  1500       1
18    1.0  1900       0

Now the data looks exactly like the input data. We got a bit lucky here as the variables remain in the same order. We need not change anything else. 

In [29]:
def confusion_matrix(data,actual_values,model):
        
        # Confusion matrix 
        
        # There are three parameters
        # ----------
        # data: data frame or array
            # data is a data frame formatted in the same way as your input data (without the actual values)
            # e.g. const, var1, var2, etc. Order is very important!
        # actual_values: data frame or array
            # These are the actual values from the test_data
            # In the case of a logistic regression, it should be a single column with 0s and 1s
            
        # model: a LogitResults object
            # this is the variable where you have the fitted model 
            # e.g. results_log in this course
        # ----------
        
        #Predict the values using the Logit model
        pred_values = model.predict(data)
        # Specify the bins 
        bins=np.array([0,0.5,1])
        # Create a histogram, where if values are between 0 and 0.5 tell will be considered 0
        # if they are between 0.5 and 1, they will be considered 1
        cm = np.histogram2d(actual_values, pred_values, bins=bins)[0]
        # Calculate the accuracy
        accuracy = (cm[0,0]+cm[1,1])/cm.sum()
        # Return the confusion matrix and the accuracy
        return cm, accuracy

In [30]:
# Create a confusion matrix with the test data
cm = confusion_matrix(test_data,test_actual,results_log)
cm

(array([[5.00, 1.00],
        [1.00, 12.00]]),
 0.8947368421052632)

We see that the test data accuracy is 89.4%. Our train accuracy will always be higher than the test accuracy, as seen, because of overfitting. The regression fitted the training data as well as possible, but that does not mean that the prediction is true for all the values of the population. That's why we test data the model has never seen before. 

In [31]:
# Format for easier understanding (not needed later on)
cm_df = pd.DataFrame(cm[0])
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index={0: 'Actual 0',1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0          5.0          1.0
Actual 1          1.0         12.0

In [32]:
# Check the missclassification rate
# Note that Accuracy + Missclassification rate = 1 = 100%
print ('Missclassification rate: '+str((1+1)/19))

Missclassification rate: 0.10526315789473684


The opposite of accuracy is the Misclassification Rate = Number of Misclassified Elements / All the elements. 
Here, it is (1+1)/19 = 10.53%. 